src/lib.rs

```diff
@@ -373,3 +402,105 @@ impl Iterator for AllUsers {
         self.iter.next()
     }
 }
+
+/// An iterator over all groups on the system
+///
+/// This struct is generally created by calling [`all_groups`](fn.all_groups.html)
+pub struct AllGroups {
+    iter: std::vec::IntoIter<Group>
+}
+
+impl AllGroups {
+    pub fn new() -> AllGroups {
+        let groups = Group::parse_file(GROUP_FILE).unwrap();
+        
+        AllGroups { iter: groups.into_iter() }
+    }
+}
+
+impl Iterator for AllGroups {
+    type Item = Group;
+    
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+}
+
+///Adds a group with the specified attributes to the
+///groups database (currently `/etc/groups`)
+///
+///Returns Result with error information if the operation was not successful
+///Pleasefix -> Panics if file io was unsuccessful
+//UNOPTIMIZED: This is currently requiring two iterations:
+//  one: for determine if the group already exists
+//  two: if the user calls get_unique_group_id, which iterates over the same iterator
+pub fn add_group(name: &str, gid: u32, users: &[&str]) -> Result<(), &'static str> {
+    for group in all_groups() {
+        if group.group == name || group.gid == gid {
+            return Err("group already exists")
+        }
+    }
+    
+    let mut file = OpenOptions::new()
+        .append(true)
+        .open(GROUP_FILE)
+        .unwrap();
+    
+    let gid = &gid.to_string();
+    
+    let mut attrs = vec![name, gid];
+    for i in 0..users.len() {
+        attrs.push(users[i]);
+    }
+    
+    let entry = attrs.join(";");
+    
+    file.write(entry.as_bytes()).unwrap();
+    file.write("\n".as_bytes()).unwrap();
+    Ok(())
+}
+
+//TODO: Allow for a MIN_GID and MAX_GID config file someplace
+pub fn get_unique_group_id() -> Option<u32> {
+    for gid in MIN_GID..MAX_GID {
+        let mut used = false;
+        for group in all_groups() {
+            if gid == group.gid {
+                used = true;
+                continue;
+            }
+        }
+        if used == false {
+            return Some(gid);
+        }
+    }
+    None
+}
+
+///Adds a user with the specified attributes to the
+///users database (currently `/etc/passwd`)
+///
+///Returns Result with error information if the operation was not successful
+///Pleasefix -> Panics if file io was unsuccessful
+pub fn add_user(user: &str, uid: u32, gid: u32, name: &str, home: &str, shell: &str) -> Result<(), &'static str> {
+    for _user in all_users() {
+        if _user.user == user || _user.uid == uid {
+            return Err("user already exists");
+        }
+    }
+    
+    let mut file = OpenOptions::new()
+        .append(true)
+        .open(PASSWD_FILE)
+        .unwrap();
+    
+    let uid = &uid.to_string();
+    let gid = &gid.to_string();
+    
+    let attrs = vec![user, "", uid, gid, name, home, shell];
+    let entry = attrs.join(";");
+    
+    file.write(entry.as_bytes()).unwrap();
```

Comment:
This would be .map(|err| format!("error writing passwd: {}", err)) instead of unwrap()

Reply:
I decided to opt for a `format!` to add the newline and then use a match with only one file write for the result, rather than using one of those methods. I like it better than my other solution. Once I get it to build, I'll push here.

Reply:
Cool

Reply:
Actually, just found a major issue with my "fix" for the empty field and the end of a line. Basically it's causing read to not work, and of course, breaking sudo, among other things. Working on a fix. Don't merge unless I've actually fixed it.